test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/metrics/async/AsyncClientMetricPublisherResolutionTest.java

```diff
@@ -0,0 +1,160 @@
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.metrics.async;
+
+import static org.hamcrest.Matchers.instanceOf;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyZeroInteractions;
+import static org.mockito.Mockito.when;
+import com.github.tomakehurst.wiremock.junit.WireMockRule;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Arrays;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.metrics.MetricCollection;
+import software.amazon.awssdk.metrics.MetricPublisher;
+import software.amazon.awssdk.services.protocolrestjson.ProtocolRestJsonAsyncClient;
+import software.amazon.awssdk.services.protocolrestjson.ProtocolRestJsonAsyncClientBuilder;
+import software.amazon.awssdk.services.protocolrestjson.model.ProtocolRestJsonException;
+
+@RunWith(MockitoJUnitRunner.class)
+public class AsyncClientMetricPublisherResolutionTest {
+    @Mock
+    private AwsCredentialsProvider mockCredentialsProvider;
+
+    @Rule
+    public WireMockRule wireMock = new WireMockRule(0);
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
+    private ProtocolRestJsonAsyncClient client;
+
+
+    @Before
+    public void setup() {
+        when(mockCredentialsProvider.resolveCredentials()).thenAnswer(invocation -> {
+            try {
+                Thread.sleep(100);
+            } catch (InterruptedException ie) {
+                ie.printStackTrace();
+            }
+            return AwsBasicCredentials.create("foo", "bar");
+        });
+    }
+
+    @After
+    public void teardown() {
+        wireMock.resetAll();
+        if (client != null) {
+            client.close();
+        }
+        client = null;
+    }
+
+    @Test
+    public void testApiCall_noPublishersSet_noNpe() {
+        client = clientWithPublishers();
+        // This is thrown because all the requests to our wiremock are
+        // nonsense, it's just important that we don't get NPE because we
+        // don't have publishers set
+        thrown.expectCause(instanceOf(ProtocolRestJsonException.class));
+        client.allTypes().join();
+    }
+
+    @Test
+    public void testApiCall_publishersSetOnClient_clientPublishersInvoked() throws IOException {
+        MetricPublisher publisher1 = mock(MetricPublisher.class);
+        MetricPublisher publisher2 = mock(MetricPublisher.class);
+
+        client = clientWithPublishers(publisher1, publisher2);
+
+        try {
+            client.allTypes().join();
```

Comment:
minor: missing `fail("no exception thrown")`, same as other tests.

Reply:
Is this necessary? It doesn't really matter if an exception is thrown or not; the test is just just checking that the right publishers are being used based on configuration. It just happens to throw here because the mock isn't fully set up

Reply:
oh okay, I misread the test. We don't need it in this case